index.html

```diff
@@ -4,162 +4,89 @@
     <meta charset="UTF-8" />
     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
     <meta http-equiv="X-UA-Compatible" content="ie=edge" />
```

Comment:
It's better to inject bootstrap library here so that you don't get a broken page in the first glance.

Reply:
I know what you mean, but is there a way to achieve that? 
Since we are using `sass-loader`, I thought the only way it can get load is by `entry js` file?

Reply:
Sorry for the confusion. I mean use webpack plugin to inject the CSS to this place. I remember there is a webpack plugin that you can extract the CSS to a file. 

The alternative way is to use a loading indicator before everything is ready. I don't really recommend this behavior though.

Reply:
https://webpack.docschina.org/plugins/mini-css-extract-plugin/#extracting-all-css-in-a-single-file

Reply:
It did work but I wonder how it works?

I did some experiment on `style-loader` and find `html` will get loaded first, then `bundle.js` will get loaded in tail of html. And css is somehow include in `bundle.js`. Therefore, css will load after html.

![image](https://user-images.githubusercontent.com/13488449/61842572-0f0c1900-aedc-11e9-85ad-7aaaf0fe916e.png)

![image](https://user-images.githubusercontent.com/13488449/61842856-01a35e80-aedd-11e9-9a06-11a9ee414ed4.png)

---

With `MiniCssExtractPlugin`, css will get injected to head, like you said.

![image](https://user-images.githubusercontent.com/13488449/61842913-2f88a300-aedd-11e9-9408-536f7b592639.png)

Reply:
It seems like a default behavior that all css will merge into one css file.
Even when I am not using `optimization.splitChunks.cacheGroups` 
![image](https://user-images.githubusercontent.com/13488449/61843209-382da900-aede-11e9-90f4-ad5dcb721124.png)

Reply:
According to [here](https://webpack.docschina.org/plugins/mini-css-extract-plugin/#extracting-all-css-in-a-single-file)
![image](https://user-images.githubusercontent.com/13488449/61843315-85aa1600-aede-11e9-97bb-bf6511d2e6f9.png)

By experiment, I still have HMR behavior even with `MiniCssExtractPlugin`.
Therefore I would you `MiniCssExtractPlugin` in dev/prod env

Reply:
How does it work? Not sure if the following explanation is what you want. 

The `index.html` is processed through the webpack compilation. It's by nature you can do anything to get desired HTML result. 

`MiniCssExtractPlugin` by its name, is **CSS** and **Extraction**. The naming explained quite well that all your CSS will be extracted to somewhere. I remember I saw this setting in your webpack config. Why did you use it then?

Reply:
I found `MiniCssExtractPlugin` in an article, it says in `production env` we should use it. 
I thought it would somehow minify it since its name start with `Mini`.

Reply:
For minification and obfuscation, it's common to use `uglify` or `terser`. The latter is more popular nowadays. 

I have to use `ExtractTextPlugin` for the CSS extraction in webpack 3. The `MiniCssExtractPlugin`  only focuses on CSS in webpack 4. Maybe it means **minimal**, not minify.

Checked its [documentation](https://github.com/webpack-contrib/mini-css-extract-plugin#minimizing-for-production), it seems not containing any minification feature itself.